File: src/876_middle_of_the_linked_list/middle_of_the_linked_list.py
from typing import Optional, List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow


def build_list(arr: List[int]) -> Optional[ListNode]:
    dummy = ListNode(0)
    curr = dummy
    for x in arr:
        curr.next = ListNode(x)
        curr = curr.next
    return dummy.next


def list_to_array(head: Optional[ListNode]) -> List[int]:
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res


# Test scaffolding
if __name__ == "__main__":
    tests = [
        ([1, 2, 3, 4, 5], [3, 4, 5]),
        ([1, 2, 3, 4, 5, 6], [4, 5, 6]),
        ([1], [1]),
        ([1, 2], [2]),
        ([1, 2, 3], [2, 3]),
    ]
    sol = Solution()
    for arr, expected in tests:
        head = build_list(arr)
        actual = list_to_array(sol.middleNode(head))
        print(f"{actual} | {expected}")

File: src/876_middle_of_the_linked_list/MiddleOfTheLinkedList.java
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }
}

public class MiddleOfTheLinkedList {
    // Helper to build list from array
    public static ListNode buildList(int[] arr) {
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        for (int x : arr) {
            curr.next = new ListNode(x);
            curr = curr.next;
        }
        return dummy.next;
    }

    // Helper to convert list to array
    public static int[] listToArray(ListNode head) {
        java.util.List<Integer> res = new java.util.ArrayList<>();
        while (head != null) {
            res.add(head.val);
            head = head.next;
        }
        return res.stream().mapToInt(i -> i).toArray();
    }

    // Test scaffolding
    public static void main(String[] args) {
        int[][] tests = {
                { 1, 2, 3, 4, 5 },
                { 1, 2, 3, 4, 5, 6 },
                { 1 },
                { 1, 2 },
                { 1, 2, 3 }
        };
        int[][] expected = {
                { 3, 4, 5 },
                { 4, 5, 6 },
                { 1 },
                { 2 },
                { 2, 3 }
        };
        Solution sol = new Solution();
        for (int i = 0; i < tests.length; i++) {
            ListNode head = buildList(tests[i]);
            int[] actual = listToArray(sol.middleNode(head));
            System.out.println(java.util.Arrays.toString(actual) + " | " + java.util.Arrays.toString(expected[i]));
        }
    }
}
